File: cmd/robot/actions.go
// Copyright (C) 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"context"
	"flag"

	"github.com/google/gapid/core/app"
	"github.com/google/gapid/core/net/grpcutil"
	"github.com/google/gapid/test/robot/master"
	"google.golang.org/grpc"
)

type RobotOptions struct {
	ServerAddress string `help:"The master server address"`
}

var (
	startVerb = app.AddVerb(&app.Verb{
		Name:      "start",
		ShortHelp: "Start a server",
	})
	searchVerb = app.AddVerb(&app.Verb{
		Name:      "search",
		ShortHelp: "Search for content in the server",
	})
	uploadVerb = app.AddVerb(&app.Verb{
		Name:      "upload",
		ShortHelp: "Upload a file to a server",
	})
	setVerb = app.AddVerb(&app.Verb{
		Name:      "set",
		ShortHelp: "Sets a value in a server",
	})
	defaultRobotOptions = RobotOptions{ServerAddress: defaultMasterAddress}
)

func init() {
	app.AddVerb(&app.Verb{
		Name:      "stop",
		ShortHelp: "Stop a server",
		Action:    &stopVerb{RobotOptions: defaultRobotOptions},
	})
	app.AddVerb(&app.Verb{
		Name:      "restart",
		ShortHelp: "Restart a server",
		Action:    &restartVerb{RobotOptions: defaultRobotOptions},
	})
}

type stopVerb struct {
	RobotOptions

	Now bool `help:"Immediate shutdown"`
}

func (v *stopVerb) Run(ctx context.Context, flags flag.FlagSet) error {
	return grpcutil.Client(ctx, v.ServerAddress, func(ctx context.Context, conn *grpc.ClientConn) error {
		client := master.NewClient(ctx, master.NewRemoteMaster(ctx, conn))
		if v.Now {
			return client.Kill(ctx, flags.Args()...)
		}
		return client.Shutdown(ctx, flags.Args()...)
	}, grpc.WithInsecure())
}

type restartVerb struct {
	RobotOptions
}

func (v *restartVerb) Run(ctx context.Context, flags flag.FlagSet) error {
	return grpcutil.Client(ctx, v.ServerAddress, func(ctx context.Context, conn *grpc.ClientConn) error {
		client := master.NewClient(ctx, master.NewRemoteMaster(ctx, conn))
		return client.Restart(ctx, flags.Args()...)
	}, grpc.WithInsecure())
}

File: test/robot/web/www/js/actions.js
/*
 * Copyright (C) 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";

// The actions module provides base functionality to access the action stashes.
define(["queriable"],
function(queriable) {
  const JOB_STATUS = {
    SCHEDULED: 0,
    RUNNING: 1,
    SUCCESSFUL: 2,
    FAILED: 3
  };

  class Action extends queriable.Base {
    constructor(json) {
      super(json);
    }

    get status() {
      return this.json_.status || 0;
    }

    get statusString() {
      switch (this.status) {
        case JOB_STATUS.SCHEDULED: return "scheduled";
        case JOB_STATUS.RUNNING: return "running";
        case JOB_STATUS.SUCCESSFUL: return "successful";
        case JOB_STATUS.FAILED: return "failed";
      }
    }

    get successful() {
      return this.status == JOB_STATUS.SUCCESSFUL;
    }

    get failed() {
      return this.status == JOB_STATUS.FAILED;
    }

    get input_() {
      return this.json_.input || {};
    }

    get package() {
      return this.json_.package || {};
    }

    get output_() {
      return this.json_.output || {};
    }

    get log() {
      return this.output_.log || "";
    }

    get err() {
      return this.output_.err || "";
    }
  }

  return {
    JOB_STATUS: JOB_STATUS,
    Action: Action,
  }
});
